Use aligned_alloc directly for grpc_core::Arena

arjunroy opened the pull request.

arjunroy commented:
name                                          old time/op             new time/op             delta
BM_Arena_NoOp/1                                20.2ns ± 0%             12.7ns ± 0%  -37.04%        (p=0.000 n=16+18)
BM_Arena_NoOp/8                                20.3ns ± 5%             12.7ns ± 0%  -37.27%        (p=0.000 n=16+17)
BM_Arena_NoOp/64                               20.6ns ± 0%             12.8ns ± 0%  -37.96%        (p=0.000 n=20+17)
BM_Arena_NoOp/512                              20.6ns ± 0%             12.8ns ± 0%  -37.81%        (p=0.000 n=20+18)
BM_Arena_NoOp/4096                             21.1ns ± 1%             13.7ns ± 0%  -35.20%        (p=0.000 n=18+17)
BM_Arena_NoOp/32768                            27.9ns ± 0%             20.9ns ± 0%  -24.92%        (p=0.000 n=19+17)
BM_Arena_NoOp/262144                            126ns ± 2%              123ns ± 2%   -2.82%        (p=0.000 n=20+20)
BM_Arena_NoOp/1048576                           178ns ± 1%              173ns ± 1%   -2.76%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/1/1                         43.3ns ± 1%             35.2ns ± 1%  -18.79%        (p=0.000 n=19+20)
BM_Arena_ManyAlloc/8/1                         39.3ns ± 3%             33.9ns ± 6%  -13.67%        (p=0.000 n=19+20)
BM_Arena_ManyAlloc/64/1                        38.3ns ± 3%             33.6ns ± 5%  -12.41%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/512/1                       38.5ns ± 4%             34.0ns ± 4%  -11.71%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/4096/1                      38.3ns ± 2%             34.1ns ± 4%  -10.89%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/32768/1                     43.5ns ± 2%             34.1ns ± 4%  -21.56%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/262144/1                    43.8ns ± 3%             33.5ns ± 5%  -23.39%        (p=0.000 n=19+20)
BM_Arena_ManyAlloc/1048576/1                   43.5ns ± 2%             33.6ns ± 6%  -22.75%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/1/8                         40.5ns ±11%             33.9ns ± 4%  -16.24%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/8/8                         40.4ns ±11%             33.8ns ± 6%  -16.28%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/64/8                        38.6ns ± 2%             34.1ns ± 4%  -11.58%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/512/8                       38.4ns ± 2%             34.3ns ± 4%  -10.73%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/4096/8                      38.6ns ± 2%             33.8ns ± 5%  -12.37%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/32768/8                     43.5ns ± 1%             33.8ns ± 4%  -22.20%        (p=0.000 n=19+20)
BM_Arena_ManyAlloc/262144/8                    43.7ns ± 2%             34.2ns ± 4%  -21.83%        (p=0.000 n=19+20)
BM_Arena_ManyAlloc/1048576/8                   43.5ns ± 1%             33.8ns ± 5%  -22.36%        (p=0.000 n=17+20)
BM_Arena_ManyAlloc/1/64                        61.9ns ± 2%             74.7ns ± 2%  +20.76%        (p=0.000 n=19+19)
BM_Arena_ManyAlloc/8/64                        62.3ns ± 2%             75.3ns ± 2%  +20.90%        (p=0.000 n=18+20)
BM_Arena_ManyAlloc/64/64                       62.0ns ± 1%             75.4ns ± 2%  +21.72%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/512/64                      62.5ns ± 1%             74.5ns ± 2%  +19.21%        (p=0.000 n=17+17)
BM_Arena_ManyAlloc/4096/64                     62.1ns ± 1%             74.9ns ± 4%  +20.76%        (p=0.000 n=19+19)
BM_Arena_ManyAlloc/32768/64                    62.2ns ± 2%             74.9ns ± 2%  +20.32%        (p=0.000 n=20+18)
BM_Arena_ManyAlloc/262144/64                   62.1ns ± 2%             75.1ns ± 3%  +20.97%        (p=0.000 n=19+19)
BM_Arena_ManyAlloc/1048576/64                  62.0ns ± 1%             74.9ns ± 3%  +20.81%        (p=0.000 n=20+16)
BM_Arena_ManyAlloc/1/512                        180ns ± 3%              164ns ± 3%   -8.89%        (p=0.000 n=18+20)
BM_Arena_ManyAlloc/8/512                        180ns ± 3%              164ns ± 3%   -8.56%        (p=0.000 n=18+20)
BM_Arena_ManyAlloc/64/512                       179ns ± 3%              165ns ± 3%   -7.88%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/512/512                      179ns ± 3%              168ns ± 2%   -5.74%        (p=0.000 n=19+19)
BM_Arena_ManyAlloc/4096/512                     181ns ± 2%              164ns ± 3%   -9.27%        (p=0.000 n=19+20)
BM_Arena_ManyAlloc/32768/512                    181ns ± 3%              165ns ± 2%   -8.84%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/262144/512                   180ns ± 3%              165ns ± 2%   -8.32%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/1048576/512                  181ns ± 3%              165ns ± 2%   -8.96%        (p=0.000 n=20+20)
BM_Arena_ManyAlloc/1/32768                      591ns ±10%              564ns ± 4%   -4.47%        (p=0.030 n=20+20)
BM_Arena_Batch/1/1/1                           26.6ns ± 0%             23.8ns ± 0%  -10.75%        (p=0.000 n=17+19)
BM_Arena_Batch/8/1/1                           26.6ns ± 0%             23.8ns ± 0%  -10.75%        (p=0.000 n=17+20)
BM_Arena_Batch/64/1/1                          26.3ns ± 0%             23.8ns ± 0%   -9.48%        (p=0.000 n=16+18)
BM_Arena_Batch/512/1/1                         26.4ns ± 0%             23.9ns ± 1%   -9.31%        (p=0.000 n=19+20)
BM_Arena_Batch/4096/1/1                        27.6ns ± 1%             24.6ns ± 1%  -10.70%        (p=0.000 n=19+18)
BM_Arena_Batch/32768/1/1                       34.1ns ± 1%             31.7ns ± 3%   -6.91%        (p=0.000 n=18+20)
BM_Arena_Batch/65536/1/1                       41.4ns ± 2%             39.0ns ± 4%   -5.83%        (p=0.000 n=19+20)
BM_Arena_Batch/1/8/1                            229ns ± 9%              198ns ± 0%  -13.91%        (p=0.000 n=20+16)
BM_Arena_Batch/8/8/1                            229ns ± 9%              198ns ± 0%  -13.88%        (p=0.000 n=20+16)
BM_Arena_Batch/64/8/1                           164ns ± 0%              150ns ± 0%   -8.85%        (p=0.000 n=15+17)
BM_Arena_Batch/512/8/1                         90.9ns ± 0%             88.7ns ± 0%   -2.47%        (p=0.000 n=16+17)
BM_Arena_Batch/4096/8/1                        91.8ns ± 0%             89.2ns ± 0%   -2.85%        (p=0.000 n=18+19)
BM_Arena_Batch/32768/8/1                       98.5ns ± 1%             95.8ns ± 0%   -2.73%        (p=0.000 n=18+20)
BM_Arena_Batch/65536/8/1                        106ns ± 2%              103ns ± 1%   -2.83%        (p=0.000 n=20+20)
BM_Arena_Batch/1/64/1                         1.89µs ± 9%             1.62µs ± 0%  -14.11%        (p=0.000 n=20+16)
BM_Arena_Batch/8/64/1                         1.89µs ± 9%             1.62µs ± 0%  -14.10%        (p=0.000 n=20+16)
BM_Arena_Batch/64/64/1                        1.76µs ± 0%             1.57µs ± 0%  -10.33%        (p=0.000 n=18+17)
BM_Arena_Batch/512/64/1                       1.22µs ± 1%             1.12µs ± 0%   -7.94%        (p=0.000 n=18+16)
BM_Arena_Batch/4096/64/1                        611ns ± 0%              608ns ± 0%   -0.42%        (p=0.000 n=19+16)
BM_Arena_Batch/32768/64/1                       618ns ± 0%              615ns ± 0%   -0.43%        (p=0.000 n=20+20)
BM_Arena_Batch/65536/64/1                       625ns ± 0%              622ns ± 0%   -0.45%        (p=0.000 n=20+20)
BM_Arena_Batch/1/1/8                           26.7ns ± 0%             23.8ns ± 0%  -10.78%        (p=0.000 n=20+20)
BM_Arena_Batch/8/1/8                           26.7ns ± 0%             23.8ns ± 0%  -10.82%        (p=0.000 n=20+19)
BM_Arena_Batch/64/1/8                          26.3ns ± 0%             23.8ns ± 0%   -9.56%        (p=0.000 n=19+18)
BM_Arena_Batch/512/1/8                         26.3ns ± 0%             23.9ns ± 0%   -9.43%        (p=0.000 n=17+18)
BM_Arena_Batch/4096/1/8                        27.2ns ± 1%             24.6ns ± 0%   -9.66%        (p=0.000 n=19+19)
BM_Arena_Batch/32768/1/8                       33.9ns ± 1%             31.3ns ± 2%   -7.48%        (p=0.000 n=19+20)
BM_Arena_Batch/65536/1/8                       41.1ns ± 2%             38.7ns ± 2%   -5.98%        (p=0.000 n=20+19)
BM_Arena_Batch/1/8/8                            232ns ± 8%              198ns ± 2%  -14.77%        (p=0.000 n=20+17)
BM_Arena_Batch/8/8/8                            232ns ± 8%              198ns ± 3%  -14.61%        (p=0.000 n=20+18)
BM_Arena_Batch/64/8/8                           164ns ± 1%              150ns ± 2%   -8.55%        (p=0.000 n=17+20)
BM_Arena_Batch/512/8/8                         90.9ns ± 0%             88.7ns ± 0%   -2.39%        (p=0.000 n=16+20)
BM_Arena_Batch/4096/8/8                        91.8ns ± 0%             89.2ns ± 0%   -2.76%        (p=0.000 n=19+19)
BM_Arena_Batch/32768/8/8                       98.5ns ± 0%             95.8ns ± 1%   -2.75%        (p=0.000 n=19+20)
BM_Arena_Batch/65536/8/8                        106ns ± 1%              103ns ± 1%   -2.59%        (p=0.000 n=20+19)
BM_Arena_Batch/1/64/8                         1.90µs ±10%             1.62µs ± 1%  -14.52%        (p=0.000 n=20+16)
BM_Arena_Batch/8/64/8                         1.90µs ±10%             1.62µs ± 1%  -14.43%        (p=0.000 n=20+16)
BM_Arena_Batch/64/64/8                        1.76µs ± 1%             1.58µs ± 2%  -10.01%        (p=0.000 n=18+18)
BM_Arena_Batch/512/64/8                       1.22µs ± 1%             1.13µs ± 2%   -7.54%        (p=0.000 n=16+18)
BM_Arena_Batch/4096/64/8                        611ns ± 0%              608ns ± 0%   -0.42%        (p=0.000 n=16+17)
BM_Arena_Batch/32768/64/8                       618ns ± 0%              615ns ± 0%   -0.43%        (p=0.000 n=19+19)
BM_Arena_Batch/65536/64/8                       625ns ± 0%              622ns ± 0%   -0.44%        (p=0.000 n=20+20)
BM_Arena_Batch/1/1/64                          42.7ns ± 1%             36.9ns ± 0%  -13.60%        (p=0.000 n=19+18)
BM_Arena_Batch/8/1/64                          42.6ns ± 1%             36.8ns ± 0%  -13.61%        (p=0.000 n=19+17)
BM_Arena_Batch/64/1/64                         26.3ns ± 0%             23.8ns ± 0%   -9.53%        (p=0.000 n=19+18)
BM_Arena_Batch/512/1/64                        26.3ns ± 0%             23.9ns ± 1%   -9.30%        (p=0.000 n=17+20)
BM_Arena_Batch/4096/1/64                       27.3ns ± 1%             24.7ns ± 1%   -9.55%        (p=0.000 n=20+20)
BM_Arena_Batch/32768/1/64                      33.9ns ± 3%             31.4ns ± 4%   -7.37%        (p=0.000 n=20+20)
BM_Arena_Batch/65536/1/64                      41.2ns ± 3%             39.0ns ± 5%   -5.34%        (p=0.000 n=20+20)
BM_Arena_Batch/1/8/64                           243ns ± 2%              214ns ± 2%  -11.62%        (p=0.000 n=20+18)
BM_Arena_Batch/8/8/64                           243ns ± 2%              214ns ± 2%  -11.69%        (p=0.000 n=20+18)
BM_Arena_Batch/64/8/64                          223ns ± 2%              198ns ± 2%  -10.85%        (p=0.000 n=18+18)
BM_Arena_Batch/512/8/64                        91.0ns ± 0%             88.7ns ± 0%   -2.43%        (p=0.000 n=18+20)
BM_Arena_Batch/4096/8/64                       91.8ns ± 0%             89.2ns ± 0%   -2.83%        (p=0.000 n=17+18)
BM_Arena_Batch/32768/8/64                      98.4ns ± 1%             95.8ns ± 1%   -2.70%        (p=0.000 n=18+20)
BM_Arena_Batch/65536/8/64                       106ns ± 1%              103ns ± 2%   -2.58%        (p=0.000 n=20+20)
BM_Arena_Batch/1/64/64                        1.84µs ± 1%             1.64µs ± 1%  -10.73%        (p=0.000 n=20+17)
BM_Arena_Batch/8/64/64                        1.84µs ± 1%             1.64µs ± 1%  -10.73%        (p=0.000 n=20+18)
BM_Arena_Batch/64/64/64                       1.81µs ± 1%             1.63µs ± 1%  -10.37%        (p=0.000 n=18+18)
BM_Arena_Batch/512/64/64                      1.68µs ± 0%             1.51µs ± 1%   -9.85%        (p=0.000 n=18+20)
BM_Arena_Batch/4096/64/64                       611ns ± 0%              608ns ± 0%   -0.42%        (p=0.000 n=16+18)
BM_Arena_Batch/32768/64/64                      617ns ± 0%              615ns ± 0%   -0.42%        (p=0.000 n=16+18)
BM_Arena_Batch/65536/64/64                      625ns ± 0%              622ns ± 0%   -0.43%        (p=0.000 n=20+20)
BM_Arena_Batch/1/1/512                         42.5ns ± 0%             36.9ns ± 1%  -13.11%        (p=0.000 n=19+19)
BM_Arena_Batch/8/1/512                         42.5ns ± 0%             36.9ns ± 0%  -13.18%        (p=0.000 n=18+18)
BM_Arena_Batch/64/1/512                        41.7ns ± 0%             36.9ns ± 0%  -11.41%        (p=0.000 n=18+16)
BM_Arena_Batch/512/1/512                       26.3ns ± 0%             23.9ns ± 1%   -9.33%        (p=0.000 n=18+19)
BM_Arena_Batch/4096/1/512                      27.3ns ± 1%             24.7ns ± 0%   -9.50%        (p=0.000 n=18+18)
BM_Arena_Batch/32768/1/512                     33.8ns ± 1%             31.5ns ± 3%   -6.81%        (p=0.000 n=19+20)
BM_Arena_Batch/65536/1/512                     41.2ns ± 2%             39.0ns ± 5%   -5.29%        (p=0.000 n=20+20)
BM_Arena_Batch/1/8/512                          239ns ± 1%              215ns ± 2%  -10.13%        (p=0.000 n=18+19)
BM_Arena_Batch/8/8/512                          239ns ± 1%              215ns ± 2%  -10.25%        (p=0.000 n=20+19)
BM_Arena_Batch/64/8/512                         239ns ± 1%              216ns ± 2%   -9.66%        (p=0.000 n=19+19)
BM_Arena_Batch/512/8/512                        220ns ± 1%              198ns ± 1%   -9.83%        (p=0.000 n=17+18)
BM_Arena_Batch/4096/8/512                      91.8ns ± 0%             89.2ns ± 0%   -2.78%        (p=0.000 n=17+19)
BM_Arena_Batch/32768/8/512                     98.3ns ± 0%             95.8ns ± 0%   -2.52%        (p=0.000 n=16+19)
BM_Arena_Batch/65536/8/512                      106ns ± 1%              103ns ± 1%   -2.58%        (p=0.000 n=20+19)
BM_Arena_Batch/1/64/512                       1.81µs ± 1%             1.65µs ± 1%   -9.11%        (p=0.000 n=18+19)
BM_Arena_Batch/8/64/512                       1.81µs ± 1%             1.65µs ± 1%   -9.11%        (p=0.000 n=18+19)
BM_Arena_Batch/64/64/512                      1.82µs ± 1%             1.65µs ± 1%   -9.25%        (p=0.000 n=18+19)
BM_Arena_Batch/512/64/512                     1.80µs ± 1%             1.63µs ± 0%   -9.46%        (p=0.000 n=17+16)
BM_Arena_Batch/4096/64/512                    1.69µs ± 9%             1.52µs ± 1%   -9.75%        (p=0.000 n=17+20)
BM_Arena_Batch/32768/64/512                     617ns ± 0%              615ns ± 0%   -0.39%        (p=0.000 n=18+19)
BM_Arena_Batch/65536/64/512                     625ns ± 0%              622ns ± 0%   -0.46%        (p=0.000 n=20+19)
BM_Arena_Batch/1/1/1024                        42.9ns ± 0%             37.5ns ± 0%  -12.72%        (p=0.000 n=20+19)
BM_Arena_Batch/8/1/1024                        42.9ns ± 0%             37.5ns ± 0%  -12.72%        (p=0.000 n=19+19)
BM_Arena_Batch/64/1/1024                       42.5ns ± 0%             37.5ns ± 1%  -11.75%        (p=0.000 n=17+18)
BM_Arena_Batch/512/1/1024                      42.5ns ± 0%             37.5ns ± 0%  -11.80%        (p=0.000 n=16+16)
BM_Arena_Batch/4096/1/1024                     27.3ns ± 0%             24.7ns ± 0%   -9.53%        (p=0.000 n=17+15)
BM_Arena_Batch/32768/1/1024                    33.7ns ± 0%             31.5ns ± 3%   -6.47%        (p=0.000 n=18+19)
BM_Arena_Batch/65536/1/1024                    41.3ns ± 5%             39.3ns ± 5%   -4.92%        (p=0.000 n=20+19)
BM_Arena_Batch/1/8/1024                         243ns ± 1%              219ns ± 3%  -10.18%        (p=0.000 n=19+19)
BM_Arena_Batch/8/8/1024                         243ns ± 1%              219ns ± 3%  -10.16%        (p=0.000 n=19+19)
BM_Arena_Batch/64/8/1024                        243ns ± 1%              219ns ± 1%   -9.72%        (p=0.000 n=18+19)
BM_Arena_Batch/512/8/1024                       242ns ± 1%              217ns ± 1%  -10.08%        (p=0.000 n=18+16)
BM_Arena_Batch/4096/8/1024                      166ns ± 0%              154ns ± 0%   -7.52%        (p=0.000 n=16+18)
BM_Arena_Batch/32768/8/1024                    98.3ns ± 0%             95.8ns ± 1%   -2.55%        (p=0.000 n=19+19)
BM_Arena_Batch/65536/8/1024                     106ns ± 1%              103ns ± 2%   -2.25%        (p=0.000 n=20+19)
BM_Arena_Batch/1/64/1024                      1.84µs ± 0%             1.67µs ± 1%   -9.19%        (p=0.000 n=17+19)
BM_Arena_Batch/8/64/1024                      1.84µs ± 0%             1.67µs ± 1%   -9.23%        (p=0.000 n=18+19)
BM_Arena_Batch/64/64/1024                     1.84µs ± 1%             1.67µs ± 1%   -9.28%        (p=0.000 n=18+18)
BM_Arena_Batch/512/64/1024                    1.84µs ± 0%             1.67µs ± 0%   -9.23%        (p=0.000 n=18+19)
BM_Arena_Batch/4096/64/1024                   1.77µs ± 0%             1.61µs ± 0%   -8.91%        (p=0.000 n=17+19)
BM_Arena_Batch/32768/64/1024                  1.23µs ± 0%             1.15µs ± 0%   -7.11%        (p=0.000 n=19+19)
BM_Arena_Batch/65536/64/1024                    625ns ± 0%              622ns ± 0%   -0.43%        (p=0.000 n=20+19)

markdroth commented on src/core/lib/gprpp/arena.cc:
```diff
@@ -77,7 +96,9 @@ void* Arena::AllocZone(size_t size) {
   size_t alloc_size = zone_base_size + size;
-  Zone* z = new (gpr_arena_malloc<GPR_MAX_ALIGNMENT>(alloc_size)) Zone();
+  void* storage = _aligned_alloc<GPR_MAX_ALIGNMENT>(alloc_size);
+  GPR_ASSERT(storage != nullptr);
```
Is it actually possible for this to be null?  Linux does allocate-on-write memory management, so most of our code is written to assume that mallocs can never fail.

arjunroy replied:
We assume that malloc doesn't fail for the returned pointer because we call abort() if the returned pointer was null - so our code never has to deal with it since we already took the process out to pasture.

This check here is the equivalent of the abort() we do in the existing alloc code.

markdroth commented on src/core/lib/gprpp/arena.cc:
```diff
@@ -31,19 +31,36 @@
 template <size_t alignment>
-static void* gpr_arena_malloc(size_t size) {
+static void* _aligned_alloc(size_t size) {
+#if defined(GPR_LINUX)
```
Is `aligned_alloc()` only available on Linux?  If it could be available on other platforms, then it might be better to add a new `GPR_HAS_ALIGNED_ALLOC` macro to port_platform.h and use that to decide whether to use it.

arjunroy replied:
aligned_alloc is available on Linux in c++11 in stdlib.h.
std::aligned_alloc is available in c++17 in <cstdlib>.
It looks like Windows gives _aligned_alloc.
For osx I think the solution is to use posix_memalign.

Then, for all other platforms, for now we can devolve to our implementation.

arjunroy replied:
Added as requested.

markdroth commented on src/core/lib/gprpp/arena.cc:
```diff
@@ -31,19 +31,36 @@
 template <size_t alignment>
-static void* gpr_arena_malloc(size_t size) {
+static void* _aligned_alloc(size_t size) {
+#if defined(GPR_LINUX)
```
Instead of having this compile-time switch here, how about doing it in the definition of `gpr_malloc_aligned()` and `gpr_free_aligned()`?  That way, any code that calls those functions can get this benefit.

arjunroy replied:
Right now the only callers are the grpc_call arena, and the global (initialization time allocated) handshaker list. No other users at present.

My preference is to keep it here since:
1) I already changed gpr_malloc_aligned/gpr_free_aligned to use this also anyways,
2) If we use the function pointers method as you and Soheil spoke about, a call to arena destruction would do:
- a call to perform gpr_free_aligned() which does a jump to a runtime-initialized function pointer that happens to be aligned_alloc (or whatever alternative) (2 calls)
- on non windows/linux/posix platforms, that alternative would be our in house free_aligned() which does another jump to gpr_free and then another jump to free (4 calls)

Rather, this way it's a single call to a statically known target. I'm not sure what the non-statically intialized global function pointer table actually buys us.

arjunroy replied:
Addressed in latest PR.

soheilhy commented:
Thank you @arjunroy  for taking a stab at this. IMHO, the correct implementation for this would be:

1) Add two callbacks called `aligned_alloc_fn` and `aligned_free_fn` in `gpr_allocation_functions` in `include/grpc/support/alloc.h`.
2) Then, add the proper implement based on the platform to `g_alloc_functions` with proper macro definition in <port_platforms.h>. the macros should be called something like `GPR_HAS_ALIGNED_ALLOC`.
3) Then, change the implementation of `gpr_malloc_aligned` and  `gpr_free_aligned` to use those callbacks.

arjunroy commented:
Updated PR based on comments from @roth and @soheilhy . Specifically:
- Added port_platform.h definitions for whether we have aligned_alloc (linux), _aligned_malloc(windows) and posix_memalign (osx) functions available.
- Added function pointers for malloc_aligned, malloc_cacheline and free_aligned to pointer structure.
- Made it so that pointers are changeable on init if desired.

Specifically, by default, we use the gpr_malloc_aligned/gpr_malloc_cacheline/gpr_free_aligned structures, which use the port_platform definitions to choose what underlying function to call. There is a fallback to the old gpr_malloc based implementation if all else fails.

If we set at init time custom allocators, we ensure that either malloc_align/malloc_cache/free_align are all provided, or all null. If all null, we fallback to using the old method that is based on gpr_malloc.

soheilhy commented on include/grpc/support/alloc.h:
```diff
@@ -32,6 +32,9 @@ typedef struct gpr_allocation_functions {
   void* (*realloc_fn)(void* ptr, size_t size);
   void (*free_fn)(void* ptr);
+  void* (*aligned_alloc_fn)(size_t size, size_t alignment);
+  void* (*cacheline_alloc_fn)(size_t size);
```
What is the significance of `cacheline_alloc`? That basically is `aligned_alloc_fn(CACHE_LINE)`?

arjunroy replied:
Removed.

soheilhy commented on src/core/lib/gpr/alloc.cc:
```diff
@@ -47,6 +83,15 @@ void gpr_set_allocation_functions(gpr_allocation_functions functions) {
+  GPR_ASSERT((functions.cacheline_alloc_fn == nullptr) ==
+             (functions.aligned_free_fn == nullptr));
+  if (functions.aligned_alloc_fn == nullptr) {
+    functions.aligned_alloc_fn = aligned_alloc_with_gpr_malloc;
```
This should be `gpr_malloc_aligned`?
Same for the rest of them.

soheilhy replied:
This version of the PR is different than what I have seen initially. I think we should use `platform_malloc_aligned` here.

My point is that `platform_malloc_aligned` is the default and `aligned_alloc_with_gpr_malloc` is a fallback that it uses. Otherwise, you'll make old binaries that are not aware of your nice change, just slower than they should be.

arjunroy replied:
Done.

soheilhy commented on src/core/lib/gpr/alloc.cc:
```diff
@@ -88,12 +133,58 @@ void* gpr_realloc(void* p, size_t size) {
 }
 
-void gpr_free_aligned(void* ptr) { gpr_free((static_cast<void**>(ptr))[-1]); }
+void* gpr_malloc_cacheline(size_t size) {
```
related to my comment above. I think this is not really required and should be removed. users can use aligned-alloc with cache line size.

soheilhy commented on src/core/lib/gprpp/arena.cc:
```diff
@@ -58,7 +53,10 @@ Arena* Arena::Create(size_t initial_size) {
           ? GPR_CACHELINE_SIZE
           : GPR_MAX_ALIGNMENT;
-  return new (gpr_arena_malloc<alignment>(alloc_size)) Arena(initial_size);
+  void* storage = (alignment == GPR_CACHELINE_SIZE)
```
related to the same comment above: this branch is unnecessary if you just used gpr_malloc_aligned.

soheilhy commented on test/core/gpr/alloc_test.cc:
```diff
@@ -35,8 +35,8 @@ static void test_custom_allocs() {
-  gpr_allocation_functions fns = {fake_malloc, nullptr, fake_realloc,
-                                  fake_free};
+  gpr_allocation_functions fns = {fake_malloc, nullptr, fake_realloc, fake_free,
+                                  nullptr,     nullptr, nullptr};
```
please add a test when these are not nullptr.

arjunroy replied:
Done.

markdroth commented on src/core/lib/gpr/alloc.cc:
```diff
@@ -88,12 +133,58 @@ void* gpr_realloc(void* p, size_t size) {
+  GPR_ASSERT(ret != nullptr);
+  return ret;
+#elif defined(GPR_HAS_ALIGNED_MALLOC)
+  GPR_DEBUG_ASSERT(is_power_of_two(alignment));
```
Why is this assertion being done in two of the three cases here?  Shouldn't it be done in all cases?

markdroth commented on src/core/lib/gprpp/arena.cc:
```diff
@@ -77,7 +75,8 @@ void* Arena::AllocZone(size_t size) {
   size_t alloc_size = zone_base_size + size;
-  Zone* z = new (gpr_arena_malloc<GPR_MAX_ALIGNMENT>(alloc_size)) Zone();
+  void* storage = gpr_malloc_aligned(alloc_size, GPR_MAX_ALIGNMENT);
+  Zone* z = new (storage) Zone();
```
Any reason not to do this all on one line, like was being done before?

markdroth commented on src/core/lib/gpr/alloc.h:
```diff
@@ -25,4 +25,12 @@
+  (((x) + GPR_CACHELINE_SIZE - 1u) & ~(GPR_CACHELINE_SIZE - 1u))
+
+#define GPR_ROUND_UP_TO_SPECIFIED_SIZE(x, align) \
+  (((x) + align - 1u) & ~(align - 1u))
```
Suggest parens around `align`, in both occurrences.  That way, if someone passes a complex expression as the `align` parameter of the macro, it won't break the expression due to operator precedence rules.

arjunroy replied:
Done.

markdroth commented on src/core/lib/gpr/alloc.h:
```diff
@@ -25,4 +25,12 @@
 #define GPR_ROUND_UP_TO_ALIGNMENT_SIZE(x) \
   (((x) + GPR_MAX_ALIGNMENT - 1u) & ~(GPR_MAX_ALIGNMENT - 1u))
 
+#define GPR_ROUND_UP_TO_CACHELINE_SIZE(x) \
```
Maybe define `GPR_ROUND_UP_TO_ALIGNMENT_SIZE()` and `GPR_ROUND_UP_TO_CACHELINE_SIZE()` as invocations of `GPR_ROUND_UP_TO_SPECIFIED_SIZE()`?

arjunroy replied:
Done.

markdroth commented on src/core/lib/gpr/alloc.h:
```diff
@@ -25,4 +25,12 @@
+#define GPR_ROUND_UP_TO_CACHELINE_SIZE(x) \
+  (((x) + GPR_CACHELINE_SIZE - 1u) & ~(GPR_CACHELINE_SIZE - 1u))
+
+#define GPR_ROUND_UP_TO_SPECIFIED_SIZE(x, align) \
```
Maybe rename this to `GPR_ROUND_UP_TO_ALIGNMENT_SIZE()`, and rename the one on line 25 to `GPR_ROUND_UP_TO_MAX_ALIGNMENT_SIZE()`, to better reflect what each one does?

arjunroy replied:
Done.

markdroth commented on src/core/lib/gpr/alloc.h:
```diff
@@ -25,4 +25,12 @@
+#define GPR_ROUND_UP_TO_SPECIFIED_SIZE(x, align) \
+  (((x) + align - 1u) & ~(align - 1u))
+
+void* gpr_malloc_cacheline(size_t size);
```
I think this can go away now.

arjunroy replied:
Done.

markdroth commented on src/core/lib/gprpp/arena.cc:
```diff
@@ -58,7 +53,8 @@ Arena* Arena::Create(size_t initial_size) {
           : GPR_MAX_ALIGNMENT;
-  return new (gpr_arena_malloc<alignment>(alloc_size)) Arena(initial_size);
+  void* storage = gpr_malloc_aligned(alloc_size, alignment);
+  return new (storage) Arena(initial_size);
```
The two lines above could be combined again.

arjunroy replied:
Done.

markdroth commented on test/core/util/memory_counters.cc:
```diff
@@ -90,8 +90,12 @@ static void guard_free(void* vptr) {
+// NB: We do not specify guard_malloc_aligned/guard_free_aligned methods. Since
+// they are null, calls to gpr_malloc_aligned/gpr_free_aligned are executed as a
+// wrapper over gpr_malloc/gpr_free, which do use guard_malloc/guard_free, and
+// thus there allocations are tracked as well.
```
s/there/their/

arjunroy replied:
Fixed.

arjunroy commented:
MSAN flake: https://github.com/grpc/grpc/issues/17855
ASAN flake: https://github.com/grpc/grpc/issues/18805

apolcyn commented:
It looks like this PR caused gRPC's dependency on libc when under `GPR_BACKWARDS_COMPATIBILITY_MODE` to increase from `2.12` to `2.16`, because `aligned_alloc` was introduced to libc only in the 2.16 release.

Libc dependency upgrades have caused issues in the past with smaller upgrades (see https://github.com/grpc/grpc/issues/11346), so I think this needs to be fixed before the 1.21 release - cc @lidizheng . One option to fix forward may be to further avoid `aligned_alloc` when under `GPR_BACKWARDS_COMPATIBILITY_MODE`

Noticed in recent errors from the [grpc_distribtests.cfg](https://github.com/grpc/grpc/blob/master/tools/internal_ci/linux/grpc_distribtests.cfg) test suite (this is ran on master and not per-PR; also see https://github.com/grpc/grpc/issues/18905).

nicolasnoble commented:
Ow. Yes, we should refactor this PR.

nicolasnoble commented on src/core/lib/gpr/alloc.cc:
```diff
@@ -33,8 +34,56 @@ static void* zalloc_with_gpr_malloc(size_t sz) {
+  void* ret = nullptr;
+  GPR_ASSERT(posix_memalign(&ret, alignment, size) == 0);
+  return ret;
+#else
```
Please do not use this pattern. No code should ever be behind a `#else`. Instead, we should have a `#error`. The "default" case should be another `#define` that we can toggle. For instance, due to #18905, we need to be able to toggle the default case for backward compatibility purpose when shipping binary files.

arjunroy replied:
@nicolasnoble  What would you recommend instead? The high level goal here is to use a fallback method if no platform specific mode is present. https://github.com/grpc/grpc/pull/18965
